Fix #28011 - remove page number from query on layered navigation swatch filter

Bartlomiejsz opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This change makes swatch layered navigation filter consistent with standard filters, by removing `p` query parameter from filter url in same way as in https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Model/Layer/Filter/Item.php
Fixes #28011 

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#28011: ElasticSearch When applying the extra filter to the product listing, Magento keeps the same page number

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Have at least one swatch filter on layered navigation and multiple pages of products
2. Go to last page on PLP
3. Use swatch attribute as filter
4. `p` parameter should be removed from url and user should see first page, same as when non-swatch attribute is used as filter

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

Bartlomiejsz committed:
Fix #28011 - remove page number from query on layered navigation swatch filter

lbajsarowicz commented on app/code/Magento/Swatches/Block/LayeredNavigation/RenderLayered.php:
```diff
@@ -192,7 +216,7 @@ protected function getOptionViewData(FilterItem $filterItem, Option $swatchOptio
     protected function isOptionVisible(FilterItem $filterItem)
     {
-        return $this->isOptionDisabled($filterItem) && $this->isShowEmptyResults() ? false : true;
+        return !($this->isOptionDisabled($filterItem) && $this->isShowEmptyResults());
```
Nice catch! :+1:

lbajsarowicz reviewed (dismissed):
As it's an adaptation of another bugfix (made to another part of the system), I am approving that.
Thanks for cleaning up the code style there.

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

ihor-sviziev reviewed (changes requested):
Hi @Bartlomiejsz,
Could you fix the failing static tests and cover your change with any type of test? I see you already adjusted unit tests, but I can't see any change related to the bug itself.

Bartlomiejsz commented:
Hi @ihor-sviziev, indeed I only did minor adjustments to unit test because otherwise it would fail due to new dependency injected. I didn't add new unit test since this whole test file will either need to be reworked or removed for phpunit 9 because it tests mock and not real class which I think won't work anymore.
I will try to cover this change with functional test

Bartlomiejsz committed:
Add functional test

lbajsarowicz commented on app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml:
```diff
@@ -0,0 +1,103 @@
+
+        <magentoCLI command="cron:run --group=index" stepKey="runCronIndexer"/>
+
+        <amOnPage url="$$createCategory.name$$.html" stepKey="amOnCategoryPage"/>
```
There's an Action Group that uses `url_key`, not `name` of category.
I can look for it tomorrow.

lbajsarowicz commented on app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml:
```diff
@@ -0,0 +1,103 @@
+        <selectOption selector="{{AdminProductAttributesSection.attributeDropdownByCode(textSwatchAttribute.attribute_code)}}" userInput="textSwatchOption2" stepKey="selectProduct3AttributeOption"/>
+        <actionGroup ref="SaveProductFormActionGroup" stepKey="saveProduct3"/>
+
+        <magentoCLI command="cron:run --group=index" stepKey="runCronIndexer"/>
```
```suggestion
        <magentoCron groups="index" stepKey="runCronIndexer"/>
```

lbajsarowicz reviewed (changes requested):
Actually, I really like the way that you prepared the test. It's really nice. There are only minor issues, let me send you the recommended Action Group for opening category tomorrow.

Bartlomiejsz committed:
Static and functional tests fix, apply changes after review, extract navigation to next category page to action group

Bartlomiejsz commented:
Hi @lbajsarowicz, I believe I found actionGroup that you mentioned. I updated PR with changes from review, also I extracted one part to action group, added fix for issue caused by other tests not being isolated and added `block_html` and `collections` caches to be cleaned, since I noticed those cache types are also invalidated when eav attribute is created

lbajsarowicz commented on app/code/Magento/Catalog/Test/Mftf/ActionGroup/AssertStorefrontCategoryCurrentPageIsFirstActionGroup.xml:
```diff
@@ -0,0 +1,17 @@
+-->
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="AssertStorefrontCategoryCurrentPageIsFirstActionGroup">
```
How about doing that more universal by changing to StorefrontAssertCategoryCurrentPageIsNthActionGroup and pass expectedPage as an argument?

lbajsarowicz commented on app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontNavigateCategoryNextPageActionGroup.xml:
```diff
@@ -0,0 +1,15 @@
+-->
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <!-- Navigates storefront category next page from toolbar -->
```
There is <annotation> node for descriptions :-)

lbajsarowicz commented on app/code/Magento/Catalog/Test/Mftf/Section/StorefrontCategoryBottomToolbarSection.xml:
```diff
@@ -12,6 +12,6 @@
         <element name="pageNumber" type="text" selector="//*[@class='toolbar toolbar-products'][2]//a[contains(@class, 'page')]//span[2][contains(text() ,'{{var1}}')]" parameterized="true"/>
         <element name="perPage" type="select" selector="//*[@class='toolbar toolbar-products'][2]//select[@id='limiter']"/>
-        <element name="currentPage" type="text" selector=".products.wrapper + .toolbar-products .pages .current span:nth-of-type(2)"/>
+        <element name="currentPage" type="text" selector=".//*[@class='toolbar toolbar-products'][2]//li[contains(@class, 'current')]//span[2]" timeout="30"/>
```
Is there any other way - without using [n] to point the bottom pagination? Some can just change the top one and... The selector won't work for bottom one.

Bartlomiejsz replied:
I don't think there is any reliable way to do that. Changing it into `.products.wrapper + .toolbar-products` as it was before has even more chance to fail, as it's relatively easy to insert something between products and bottom toolbar, but moving or removing one of toolbars is one of things that are currently impossible - https://github.com/magento/magento2/issues/7253 🙂

lbajsarowicz commented on app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml:
```diff
@@ -0,0 +1,102 @@
+            <actionGroup ref="AddTextSwatchToProductActionGroup" stepKey="addSwatchAttribute">
+                <argument name="usedInLayeredNavigation" value="1"/>
+            </actionGroup>
+            <magentoCLI command="cache:clean block_html collections config full_page" stepKey="cleanInvalidatedCaches"/>
```
magentoCron groups=index should be enough, as it cleans cache selectively for the changed products.

lbajsarowicz reviewed (changes requested):
Now it's only about minor changes 👏

Bartlomiejsz committed:
Apply review notes

lbajsarowicz reviewed (dismissed):
I have no more comments :) Great job, thank you.

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request

ihor-sviziev reviewed (dismissed):
Good job 👍

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

_**Preconditions:**_

Sample Data;
Elasticsearch 6+;

**Steps to Reproduce:**
1. Go to Storefront;
2. Go to Men->Tops->Hoodies & Sweatshirts category (for example);
3. Go to the second **2** page;
4. On filter panel choose **black** from **Color** (for example);

**Before:**

:heavy_multiplication_x: After applying the filter, the products are not displayed

![Peek 2020-04-30 14-42](https://user-images.githubusercontent.com/51679138/80707947-79995780-8af3-11ea-9250-72dc51b4b2b3.gif)

**After:** we see filtered products on the first page

![Peek 2020-05-04 13-47](https://user-images.githubusercontent.com/51679138/80958696-1454a780-8e0e-11ea-8ec1-cf920aad1a0a.gif)

_Note:_ Tested with visual and text swatches type

slavvka reviewed (changes requested):
@Bartlomiejsz please resolve the conflicts and re-run the tests after the upgrade to PHP 7.4

Bartlomiejsz committed:
Merge branch '2.4-develop' into feature/fix_28011_page_number_on_swatch_layered_nav_filter

# Conflicts:
#	app/code/Magento/Swatches/Test/Unit/Block/LayeredNavigation/RenderLayeredTest.php

Bartlomiejsz commented:
Hi @slavvka, done

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

engcom-Charlie commented:
Failed functional test not related to changes from this PR.
